client/src/plugins/update-checks/UpdateChecks.js

```diff
@@ -0,0 +1,187 @@
+      const newerThan = latestUpdateCheckInfo ? (latestUpdateCheckInfo.latestCheckedVersion) : modelerVersion;
+      const osInfo = await config.get(OS_INFO_CONFIG_KEY);
+      const plugins = this.formatPlugins(_getGlobal('plugins').appPlugins);
+      const queryString = this.constructQueryString(editorID, modelerVersion, newerThan, osInfo, plugins);
+
+      const url = Flags.get(UPDATES_SERVER_URL) + UPDATE_CHECK_API + queryString;
+      const responseJSON = await this.sendRequest(url);
+
+      const latestVersion = responseJSON.latestVersion;
+      let newLatestUpdateCheckInfo = latestUpdateCheckInfo ? latestUpdateCheckInfo : {};
+      if (latestVersion) {
+        const downloadURL = responseJSON.downloadURL;
+        const releases = responseJSON.releases;
+        this.setState({
+          isChecking: false,
+          showModal: true,
+          latestVersionInfo: { latestVersion, downloadURL, releases },
+          currentVersion: modelerVersion
+        });
+        newLatestUpdateCheckInfo.latestCheckedVersion = latestVersion;
+      }
+
+      newLatestUpdateCheckInfo.latestUpdateTime = new Date().getTime();
+      config.set(LATEST_UPDATE_CHECK_INFO_CONFIG_KEY, newLatestUpdateCheckInfo);
+    } catch (err) {
+      this.setState({ isChecking: false, requestError: true });
+    }
+  }
+
+  appendBasicQueryParameter(queryParam, key, value) {
+    return queryParam + key + '=' + value;
+  }
+
+  constructQueryString(editorID, modelerVersion, newerThan, osInfo, plugins) {
+    let queryString = '?';
+    queryString = this.appendBasicQueryParameter(queryString, 'editorID', editorID) + '&';
+    queryString = this.appendBasicQueryParameter(queryString, 'newerThan', newerThan) + '&';
+    queryString = this.appendBasicQueryParameter(queryString, 'modelerVersion', modelerVersion) + '&';
+    queryString = this.appendBasicQueryParameter(queryString, 'os', osInfo.platform) + '&';
+    queryString = this.appendBasicQueryParameter(queryString, 'osVersion', osInfo.release);
+
+    const pluginsQueryString = plugins.reduce((totalString, plugin) => {
+      totalString += '&plugins[id]=' + plugin.id + '&plugins[name]=' + plugin.name;
+      return totalString;
+    }, '');
+
+    queryString += pluginsQueryString;
+
+    return queryString;
+  }
+
+  isTimeExceeded(previousTime) {
+    const now = new Date().getTime();
+    const hoursDiff = Math.abs(now - previousTime) / HOURS_DENOMINATOR;
+    return hoursDiff >= HOURS_LIMIT;
+  }
+
+  formatPlugins(plugins) {
+    return plugins.map((plugin) => {
+      return {
+        name: plugin.name, id: plugin.name
+      };
+    });
+  }
+
+  onClose = () => {
+    this.setState({
+      showModal: false
+    });
+  }
+
+  onGoToDownloadPage = () => {
+    const {
+      latestVersionInfo
+    } = this.state;
+
+    const {
+      _getGlobal
+    } = this.props;
+
+    _getGlobal('backend').send('external:open-url', { url: latestVersionInfo.downloadURL });
```

Comment:
Can't we open the links without `backend`?

Reply:
You are correct @barmac. Links will automatically be opened via an external tool. This is implemented [here](https://github.com/camunda/camunda-modeler/blob/develop/app/lib/index.js#L272).

Reply:
Isn't this for automatic handling of <a href=".."> links? In my code I pass an onClick event to button instead of <a href=".."/> then I go to the URL inside that method because I also want to close the modal right after opening the URL, so a simple href would not cut it. Is there a way to open an URL programatically without backend? (without hackily create an invisible <a href=".."/> and simulate a click event on it)

Reply:
Have a look at https://github.com/camunda/camunda-modeler/pull/1618/commits/12ba5d02b99f64f6982c57420d8078d746d90c61 for a basic solution. Please note that the styles are broken now.

Reply:
I'd rather not use this solution as:

1) How to (easily) render the anchor tag same as a button as we are not using bootstrap or anything? (why would I deal with copying native button CSS attributes to an anchor tag)
2) If we don't want to deal with (1) we can wrap a button with an anchor tag however it is not HTML5 valid.

Considering these two does it really worth the effort? I'd say no. Maybe I'm missing a simple solution though.

Reply:
Consider that passing `_getGlobals` to a plugin is already a technical debt. We do that only because we can't quickly prepare a proper way to pass the plugin information. Thus, whenever we remove `_getGlobals` we are going to face a problem of a link which is dressed up as a button.

Reply:
Maybe we could use a `<form>` as we clearly do some form of a submission here.

Reply:
I don't think it is a technical debt, we have not decided a solution for our in-house-plugin capabilities yet. It turns out that we do need to access some other elements more than Config which maybe implies that we need to invest some time on our plugins. in the same logic passing Config to deploymant tool plugin would be a technical debt too before we actually put it there.

So at this point:

1) Use forms -> That'd probably work but we are not submitting anything really. There are no input fields, there's no submission. So we'd use a form element just to go to an external page. (**I'd consider this a bad solution**)
2) Use anchor -> How about styling? How to make it look like a native button without some really hacky CSS trick? Would it worth it (No, I'd say.)
3) Use an hidden anchor and trigger click on it -> This is the definition of a hack (technical debt)

On the other hand we have an API called: **external:open-url** which I use here to **open** an **external** **URL**. IMO not using such APIs by doing these little tricks are the exact definition of a technical debt, but using these APIs are not.

In my honest opinion making decisions like "Returning another class from the constructor of a class because App.js is growing bigger" and "Lets not use backend to open an external URL but lets use an anchor tag instead" are not really making the codebase better. I really fail to see how your suggestions on this topic are a better solution.

Reply:
Let me jump in here with my assessment:

* Using semantic HTML elements (`<form>`) may help accessibility. We could easily improve thins in a second iteration though, if we receive the user input / UX requirement for it.

* Using `external:open-url`: Both solutions work, none is more right or wrong. Let's settle what we currently have and improve in the future if we find a clear winner.